### PR TITLE
overhaul documentation of std.string indexOf-like functions

### PR DESCRIPTION
**Pull request by wolframw**

### CODE REVIEW

File: std/string.d [hunk 1]
```diff
@@ -1914,21 +1935,23 @@ if (isSomeChar!Char && isSomeChar!Char2)
...
     Params:
-        haystack = String to search for needles in.
-        needles = Strings to search for in haystack.
-        stopIdx = slices haystack like this $(D haystack[0 .. stopIdx]). If
-            the stopIdx is greater than or equal to the length of haystack the
-            functions returns `-1`.
-        cs = Indicates whether the comparisons are case sensitive.
+        haystack = string to search needles in
+        needles = characters to search for in `haystack`
+        stopIdx = index in `haystack` to stop searching at (exclusive); defaults
```
Comment (burner) [inline std/string.d:1944]: There is no stopIdx parameter in the following function signature.
Am I missing something?

Comment (burner) [reply std/string.d:1944]: sorry I was stupid

**Comment by burner** [comment]
A quick read looks good so far.
But as documentation is not my strong suit somebody else should have a look.

File: std/string.d [hunk 2]
```diff
@@ -2097,17 +2120,28 @@ if (isSomeChar!Char && isSomeChar!Char2)
...
-            the startIdx is greater than or equal to the length of haystack the
-            functions returns `-1`.
-        cs = Indicates whether the comparisons are case sensitive.
+        haystack = string to search for needles in
+        needles = characters to search for in `haystack`
+        startIdx = index of a well-formed code point in `haystack` to start
+            searching from; defaults to 0
+        cs = specifies whether comparisons are case-sensitive
+            (`Yes.caseSensitive`) or not (`No.caseSensitive`)
+
```
Comment (dkorpel) [inline std/string.d:2133]: Trailing white space here causes CI failures

File: std/string.d [hunk 3]
```diff
@@ -901,30 +902,31 @@ private template _indexOfStr(CaseSensitive cs)
...
-        the index of the first occurrence of `sub` in `s` with
-        respect to the start index `startIdx`. If `sub` is not found,
-        then `-1` is returned.
-        If the arguments are not valid UTF, the result will still
-        be in the range [-1 .. s.length], but will not be reliable otherwise.
-        If `sub` is found the value of the returned index is at least
-        `startIdx`.
+        The index of the first occurrence of `sub` in `s`. If `sub` is not found
+        or `startIdx` is greater than or equal to `s.length`, then -1 is
+        returned. If the parameters are not valid UTF, the result will still be
```
Comment (dkorpel) [inline std/string.d:919]:
```suggestion
        returned. If the arguments are not valid UTF, the result will still be
```
I think the original was correct

File: std/string.d [hunk 4]
```diff
@@ -457,26 +457,27 @@ pure nothrow @system unittest // https://issues.dlang.org/show_bug.cgi?id=15136
 alias CaseSensitive = Flag!"caseSensitive";
 
 /++
-    Searches for character in range.
+    Searches for character `c` in string or range `s` starting at index
+    `startIdx`. If `startIdx` is not given, it defaults to 0.
```
Comment (pbackus) [inline std/string.d:461]: No need to describe each individual parameter in the [summary][1]. This part of the documentation is meant to give a brief, non-exhaustive overview that can be quickly scanned (like the first line of a good git commit message). Details like this belong in either the [description][2] or the `Params:` section.

Suggested change:
```suggestion
    Searches for a character in a string or range.
```

[1]: https://dlang.org/spec/ddoc.html#summary
[2]: https://dlang.org/spec/ddoc.html#description

File: std/string.d [hunk 5]
```diff
@@ -901,30 +902,31 @@ private template _indexOfStr(CaseSensitive cs)
 }
 
 /++
-    Searches for substring in `s`.
+    Searches for substring `sub` in string or range `s` starting at index
+    `startIdx`. If `startIdx` is not given, it defaults to 0.
```
Comment (pbackus) [inline std/string.d:906]:
```suggestion
    Searches for a substring in a string or range.
```

File: std/string.d [hunk 6]
```diff
@@ -1156,23 +1158,27 @@ unittest
 }
 
 /++
+    Searches for the last occurrence of character `c` in string `s` starting at
+    `startIdx`. If `startIdx` is not given, it defaults to 0.
```
Comment (pbackus) [inline std/string.d:1162]:
```suggestion
    Searches for the last occurrence of a character in a string.
```

File: std/string.d [hunk 7]
```diff
@@ -1345,23 +1351,31 @@ if (isSomeChar!Char)
 }
 
 /++
+    Searches for the last occurrence of substring `sub` in string `s` starting
+    at `startIdx`. If `startIdx` is not given, it defaults to 0.
```
Comment (pbackus) [inline std/string.d:1355]:
```suggestion
    Searches for the last occurrence of a substring in a string.
```

File: std/string.d [hunk 8]
```diff
@@ -1914,21 +1935,23 @@ if (isSomeChar!Char && isSomeChar!Char2)
...
 /**
-    Returns the index of the last occurrence of any of the elements in $(D
-    needles) in `haystack`. If no element of `needles` is found,
-    then `-1` is returned. The `stopIdx` slices `haystack` in the
-    following way $(D s[0 .. stopIdx]). `stopIdx` represents a codeunit
-    index in `haystack`. If the sequence ending at `startIdx` does not
-    represent a well formed codepoint, then a $(REF UTFException, std,utf) may be
-    thrown.
+    Searches the string `haystack` for the last occurrence of any of the
+    characters in `needles`.
```
Comment (pbackus) [inline std/string.d:1939]:
```suggestion
    Searches `haystack` for the last occurrence of any of the
    characters in `needles`.
```

File: std/string.d [hunk 9]
```diff
@@ -2097,17 +2120,28 @@ if (isSomeChar!Char && isSomeChar!Char2)
 }
 
 /**
-    Returns the index of the first occurrence of any character not an elements
-    in `needles` in `haystack`. If all element of `haystack` are
-    element of `needles` `-1` is returned.
+    Searches the string `haystack` for characters that are not in `needles`
+    starting at index `stardIdx`. If `startIdx` is not given, it defaults to 0.
```
Comment (pbackus) [inline std/string.d:2124]:
```suggestion
    Searches `haystack` for a character not in `needles`.
```

**Review by pbackus — DISMISSED** [review]
The first paragraph of a ddoc comment should be a concise summary—ideally a single line. See suggestions.

**Review by RazvanN7 — APPROVED** [review]
@wolframw please address the review and we are good to go.

**Review by thewilsonator — APPROVED** [review]

**Comment by wolframw** [comment]
Sorry for leaving you with my half-baked PR. I got sidetracked at some point 2 years ago and forgot about this. Thanks @LightBender and @thewilsonator for getting this merged!

**Pull request merged**